akka-remote/src/test/scala/akka/remote/artery/RemoteWatcherSpec.scala

```diff
@@ -0,0 +1,321 @@
+  val remoteAddress = remoteSystem.asInstanceOf[ExtendedActorSystem].provider.getDefaultAddress
+  def remoteAddressUid = AddressUidExtension(remoteSystem).addressUid
+
+  Seq(system, remoteSystem).foreach(muteDeadLetters(
+    akka.remote.transport.AssociationHandle.Disassociated.getClass,
+    akka.remote.transport.ActorTransportAdapter.DisassociateUnderlying.getClass)(_))
+
+  override def afterTermination() {
+    shutdown(remoteSystem)
+  }
+
+  val heartbeatRspB = HeartbeatRsp(remoteAddressUid)
+
+  def createRemoteActor(props: Props, name: String): InternalActorRef = {
+    remoteSystem.actorOf(props, name)
+    system.actorSelection(RootActorPath(remoteAddress) / "user" / name) ! Identify(name)
+    expectMsgType[ActorIdentity].ref.get.asInstanceOf[InternalActorRef]
+  }
+
+  "A RemoteWatcher" must {
+
+    "have correct interaction when watching" in {
+
+      val fd = createFailureDetector()
+      val monitorA = system.actorOf(Props[TestRemoteWatcher], "monitor1")
+      val monitorB = createRemoteActor(Props(classOf[TestActorProxy], testActor), "monitor1")
+
+      val a1 = system.actorOf(Props[MyActor], "a1").asInstanceOf[InternalActorRef]
+      val a2 = system.actorOf(Props[MyActor], "a2").asInstanceOf[InternalActorRef]
+      val b1 = createRemoteActor(Props[MyActor], "b1")
+      val b2 = createRemoteActor(Props[MyActor], "b2")
+
+      monitorA ! WatchRemote(b1, a1)
+      monitorA ! WatchRemote(b2, a1)
+      monitorA ! WatchRemote(b2, a2)
+      monitorA ! Stats
+      // (a1->b1), (a1->b2), (a2->b2)
+      expectMsg(Stats.counts(watching = 3, watchingNodes = 1))
+      expectNoMsg(100 millis)
+      monitorA ! HeartbeatTick
+      expectMsg(Heartbeat)
+      expectNoMsg(100 millis)
+      monitorA ! HeartbeatTick
+      expectMsg(Heartbeat)
+      expectNoMsg(100 millis)
+      monitorA.tell(heartbeatRspB, monitorB)
+      monitorA ! HeartbeatTick
+      expectMsg(Heartbeat)
+      expectNoMsg(100 millis)
+
+      monitorA ! UnwatchRemote(b1, a1)
+      // still (a1->b2) and (a2->b2) left
+      monitorA ! Stats
+      expectMsg(Stats.counts(watching = 2, watchingNodes = 1))
+      expectNoMsg(100 millis)
+      monitorA ! HeartbeatTick
+      expectMsg(Heartbeat)
+      expectNoMsg(100 millis)
+
+      monitorA ! UnwatchRemote(b2, a2)
+      // still (a1->b2) left
+      monitorA ! Stats
+      expectMsg(Stats.counts(watching = 1, watchingNodes = 1))
+      expectNoMsg(100 millis)
+      monitorA ! HeartbeatTick
+      expectMsg(Heartbeat)
+      expectNoMsg(100 millis)
+
+      monitorA ! UnwatchRemote(b2, a1)
+      // all unwatched
+      monitorA ! Stats
+      expectMsg(Stats.empty)
+      expectNoMsg(100 millis)
+      monitorA ! HeartbeatTick
+      expectNoMsg(100 millis)
+      monitorA ! HeartbeatTick
+      expectNoMsg(100 millis)
+
+      // make sure nothing floods over to next test
+      expectNoMsg(2 seconds)
+    }
+
+    "generate AddressTerminated when missing heartbeats" in {
```

Comment:
hmm, I am not sure. Once upon a time, explicit Quarantining did generate an AddressTerminated. It likely has been changed but my memories are hazy.